test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -99,4 +106,22 @@ static void sendDatagram(DatagramChannel dc, SocketAddress remote)
         ByteBuffer bb = ByteBuffer.wrap("Greetings!".getBytes());
         dc.send(bb, remote);
     }
+
+    static void logUnexpectedWakeup(DatagramChannel dc, Set<SelectionKey> selectedKeys) {
+        for (SelectionKey key : selectedKeys) {
+            if (!key.isValid() || !key.isReadable()) {
+                System.out.println("Invalid or unreadable key: " + key);
+                return;
```

Comment:
The loop over the selection keys and checking if valid or readable doesn't make sense here. Maybe that can be just removed. If you really want to print out the selection key then save it at L52 so that you have the SelectionKey to print.

Reply:
Since we have no idea of what is going on it's fine to loop over the set of selected keys - even if it should contain at most one entry. But if it contains a key that is either not valid or not readable we should continue after printing it, not return.